File: server/node/src/controllers/artist.ts
import { Router } from "express";
import { db } from "../app";

const artistRouter = Router();

/**
 * GET artists
 * @description Get all artists
 */
artistRouter.get("/", (req, res) => {
  const artists: string[] = [];
  db.each(
    "SELECT DISTINCT artist FROM song",
    (err, row: { artist: string }) => {
      artists.push(row.artist);
    },
    (err, count) => {
      if (err) console.error(err);

      if (err || !count) res.sendStatus(404);
      res.json(artists);
    }
  );
});

export default artistRouter;

File: server/src/server/controllers/artist.rs
use crate::{get_app_state, types::AppState};
use actix_web::{get, web, HttpResponse, Responder};
use std::sync::{Arc, Mutex};

#[get("")]
pub async fn get_artists(data: web::Data<Arc<Mutex<AppState>>>) -> impl Responder {
    let state = get_app_state!(data);
    let mut artists: Vec<String> = vec![];
    let conn = &state.db;

    match conn.prepare("SELECT DISTINCT artist FROM song") {
        Ok(mut stmt) => {
            if let Ok(iter) = stmt.query_map([], |row| {
                let artist: String = row.get(0)?;
                Ok(artist)
            }) {
                for artist in iter {
                    if let Ok(artist) = artist {
                        artists.push(artist);
                    }
                }
            }
        }
        Err(error) => {
            return HttpResponse::BadRequest().body(format!("{error}"));
        }
    }

    HttpResponse::Ok().json(artists)
}
